Mathlib/Topology/Algebra/NormedValued.lean

```diff
@@ -0,0 +1,131 @@
+    is_topological_valuation := fun U => by
+      rw [Metric.mem_nhds_iff]
+      exact ⟨fun ⟨ε, hε, h⟩  =>
+          ⟨Units.mk0 ⟨ε, le_of_lt hε⟩ (ne_of_gt hε), fun x hx ↦ h (mem_ball_zero_iff.mpr hx)⟩,
+        fun ⟨ε, hε⟩ => ⟨(ε : ℝ), NNReal.coe_pos.mpr (Units.zero_lt _),
+          fun x hx ↦ hε (mem_ball_zero_iff.mp hx)⟩⟩ }
+
+end NormedField
+
+namespace Valued
+
+variable {L : Type*} [Field L] {Γ₀ : Type*} [LinearOrderedCommGroupWithZero Γ₀]
+  [val : Valued L Γ₀] [hv : RankOne val.v]
+
+/-- The norm function determined by a rank one valuation on a field `L`. -/
+def norm_def : L → ℝ := fun x : L => hv.hom (Valued.v x)
+
+theorem norm_def_nonneg (x : L) : 0 ≤ norm_def x := by simp only [norm_def, NNReal.zero_le_coe]
+
+theorem norm_def_add_le (x y : L) : norm_def (x + y) ≤ max (norm_def x) (norm_def y) := by
+  simp only [norm_def, NNReal.coe_le_coe, le_max_iff, StrictMono.le_iff_le hv.strictMono]
+  exact le_max_iff.mp (Valuation.map_add_le_max' val.v _ _)
+
+theorem norm_def_eq_zero {x : L} (hx : norm_def x = 0) : x = 0 := by
+  simpa [norm_def, NNReal.coe_eq_zero, RankOne.hom_eq_zero_iff, zero_iff] using hx
+
+variable (L) (Γ₀)
+
+/-- The normed field structure determined by a rank one valuation. -/
+def toNormedField : NormedField L :=
+  { (inferInstance : Field L) with
+    norm := norm_def
+    dist := fun x y => norm_def (x - y)
+    dist_self := fun x => by
+      simp only [sub_self, norm_def, Valuation.map_zero, hv.hom.map_zero, NNReal.coe_zero]
+    dist_comm := fun x y => by simp only [norm_def]; rw [← neg_sub, Valuation.map_neg]
+    dist_triangle := fun x y z => by
+      simp only [← sub_add_sub_cancel x y z]
+      exact le_trans (norm_def_add_le _ _)
+        (max_le_add_of_nonneg (norm_def_nonneg _) (norm_def_nonneg _))
+    edist_dist := fun x y => by simp only [ENNReal.ofReal_eq_coe_nnreal (norm_def_nonneg _)]
+    eq_of_dist_eq_zero := fun hxy => eq_of_sub_eq_zero (norm_def_eq_zero hxy)
+    dist_eq := fun x y => rfl
+    norm_mul' := fun x y => by simp only [norm_def, ← NNReal.coe_mul, _root_.map_mul]
+    toUniformSpace := Valued.toUniformSpace
+    uniformity_dist := by
+      letI : Nonempty { ε : ℝ // ε > 0 } := nonempty_Ioi_subtype
+      ext U
+      rw [hasBasis_iff.mp (Valued.hasBasis_uniformity L Γ₀), iInf_subtype', mem_iInf_of_directed]
+      · simp only [exists_true_left, mem_principal, Subtype.exists, gt_iff_lt,
+          Subtype.coe_mk, exists_prop, true_and_iff]
+        refine ⟨fun ⟨ε, hε⟩ => ?_, fun ⟨r, hr_pos, hr⟩ => ?_⟩
+        · set δ : ℝ≥0 := hv.hom ε with hδ
+          have hδ_pos : 0 < δ := by
+            rw [hδ, ← _root_.map_zero hv.hom]
+            exact hv.strictMono (Units.zero_lt ε)
+          use δ, hδ_pos
+          apply subset_trans _ hε
+          intro x hx
+          simp only [mem_setOf_eq, norm_def, hδ, NNReal.val_eq_coe, NNReal.coe_lt_coe] at hx
+          rw [mem_setOf, ← neg_sub, Valuation.map_neg]
+          exact (RankOne.strictMono Valued.v).lt_iff_lt.mp hx
```

Comment:
These lines look very similar to 118-123, which I think are proving the other direction. Can they be factored out and can you prove both directions at once, or is this more trouble than it's worth?

Reply:
Yes, each of these blocks proves one of the inclusions. I don't see a quick way of unifying the proofs.